src/server/zset_family.cc

```diff
@@ -2840,54 +2848,108 @@ void SortIfNeeded(GeoArray* ga, Sorting sorting, uint64_t count) {
+    } else {
+      cntx->SendError(res);
+    }
+    if (write_mode) {
+      cntx->transaction->Conclude();
+    }
     return;
   }
 
   auto* rb = static_cast<RedisReplyBuilder*>(cntx->reply_builder());
   // if no matching results, the user gets an empty reply.
   if (ga.empty()) {
     rb->SendNull();
+    if (write_mode) {
+      cntx->transaction->Conclude();
+    }
     return;
   }
 
   // sort and trim by count
   SortIfNeeded(&ga, geo_ops.sorting, geo_ops.count);
 
-  // generate reply array withdist, withcoords, withhash
-  int record_size = 1;
-  if (geo_ops.withdist) {
-    record_size++;
-  }
-  if (geo_ops.withhash) {
-    record_size++;
-  }
-  if (geo_ops.withcoord) {
-    record_size++;
-  }
-  rb->StartArray(ga.size());
-  for (const auto& p : ga) {
-    // [member, dist, x, y, hash]
-    rb->StartArray(record_size);
-    rb->SendBulkString(p.member);
+  if (geo_ops.store == GeoStoreType::kNoStore) {
+    // case 1: read mode
+    // case 2: write mode, kNoStore
+    //  generate reply array withdist, withcoords, withhash
+    int record_size = 1;
     if (geo_ops.withdist) {
-      rb->SendDouble(p.dist / geo_ops.conversion);
+      record_size++;
     }
     if (geo_ops.withhash) {
-      rb->SendDouble(p.score);
+      record_size++;
     }
     if (geo_ops.withcoord) {
-      rb->StartArray(2);
-      rb->SendDouble(p.longitude);
-      rb->SendDouble(p.latitude);
+      record_size++;
+    }
+    rb->StartArray(ga.size());
+    for (const auto& p : ga) {
+      // [member, dist, x, y, hash]
+      rb->StartArray(record_size);
+      rb->SendBulkString(p.member);
+      if (geo_ops.withdist) {
+        rb->SendDouble(p.dist / geo_ops.conversion);
+      }
+      if (geo_ops.withhash) {
+        rb->SendDouble(p.score);
+      }
+      if (geo_ops.withcoord) {
+        rb->StartArray(2);
+        rb->SendDouble(p.longitude);
+        rb->SendDouble(p.latitude);
+      }
     }
+  } else {
+    // case 3: write mode, !kNoStore
+    DCHECK(geo_ops.store == GeoStoreType::kStoreDist || geo_ops.store == GeoStoreType::kStoreHash);
+    ShardId dest_shard = Shard(store_key, ga.size());
+    DVLOG(1) << "store shard:" << dest_shard << ", key " << store_key;
+    AddResult add_result;
+    vector<ScoredMemberView> smvec;
+    for (const auto& p : ga) {
+      if (geo_ops.store == GeoStoreType::kStoreDist) {
+        smvec.emplace_back(p.dist / geo_ops.conversion, p.member);
+      } else {
+        DCHECK(geo_ops.store == GeoStoreType::kStoreHash);
+        smvec.emplace_back(p.score, p.member);
+      }
+    }
+
+    auto store_cb = [&](Transaction* t, EngineShard* shard) {
+      if (shard->shard_id() == dest_shard) {
+        ZParams zparams;
+        zparams.override = true;
+        add_result =
+            OpAdd(t->GetOpArgs(shard), zparams, store_key, ScoredMemberSpan{smvec}).value();
+      }
+      return OpStatus::OK;
+    };
+    cntx->transaction->Execute(std::move(store_cb), false);
```

Comment:
this should be `Execute(.., true)` as it's the last hop of the transaction.

Reply:
When STORE is supplied, this is the last hop. Otherwise, the last hop is at line 2807.
I can either handle conclude at the end, which takes 3 lines of code, or add two more arguments, `do_store` and `already_concluded` to the whole workflow, and check them after 2807 multiple times. 
Please let me know which way do you prefer.

Reply:
FYI, STORE and STOREDIST are optional:
GEORADIUSBYMEMBER key member radius <M | KM | FT | MI> [WITHCOORD]
  [WITHDIST] [WITHHASH] [COUNT count [ANY]] [ASC | DESC] [STORE key
  | STOREDIST key]